File: frontend/src/types/legalTicket.ts
import type { LegalTicketStatus } from "./legalTicketStatus";

export const LEGAL_TICKET_TYPES = [
  "SYSTEM_ERROR",
  "QUERY_ERROR",
  "CONTACT_EXPERT",
  "REFUND_REQUEST",
] as const;

export type LegalTicketType = (typeof LEGAL_TICKET_TYPES)[number];

export const isLegalTicketType = (value: string): value is LegalTicketType =>
  LEGAL_TICKET_TYPES.some((ticketType) => ticketType === value);

export const LEGAL_TICKET_RISK_LEVELS = [
  "NONE",
  "LOW",
  "MEDIUM",
  "HIGH",
  "CRITICAL",
  "UNKNOWN",
] as const;

export type LegalTicketRiskLevel = (typeof LEGAL_TICKET_RISK_LEVELS)[number];

export const isLegalTicketRiskLevel = (value: string): value is LegalTicketRiskLevel =>
  LEGAL_TICKET_RISK_LEVELS.some((riskLevel) => riskLevel === value);

export type TicketRecipientType = "EXPERT" | "ADMIN";
export type TicketPriority = "LOW" | "NORMAL" | "HIGH" | "URGENT";
export type ConversationScope = "SELECTED_RESPONSE" | "RELATED_MESSAGES" | "FULL_CONVERSATION" | "TICKET_CONTEXT_ONLY";
export type SuggestionType =
  | "DIRECT_ANSWER"
  | "ASK_UPLOAD_CONTRACT"
  | "ASK_CONTRACT_TYPE"
  | "ASK_USER_ROLE"
  | "ASK_TARGET_CLAUSE"
  | "ASK_MORE_FACTS"
  | "SUGGEST_REVISE_CLAUSE"
  | "SUGGEST_NEGOTIATION"
  | "REDIRECT_TO_SUPPORTED_SCOPE"
  | "REFUSE_AND_REDIRECT"
  | "NONE"
  | "ASK_MORE_INFO"
  | "SUGGEST_LAWYER"
  | "REQUIRE_LAWYER";
export type UserActionHint =
  | "CONTINUE_CHAT"
  | "PROVIDE_MORE_INFO"
  | "CREATE_TICKET"
  | "UPLOAD_CONTRACT"
  | "CONTACT_LAWYER";

export interface PageResponse<T> {
  items: T[];
  page: number;
  size: number;
  totalItems: number;
  totalPages: number;
}

export interface CreateLegalTicketRequest {
  creationSource?: "AI_CHAT" | "MANUAL_FORM";
  title?: string;
  description?: string;
  recipientType?: TicketRecipientType;
  priority?: TicketPriority;
  conversationScope?: ConversationScope;
  userMessageId?: string | null;
  assistantMessageId?: string | null;
  focusedDocumentId?: string | null;
  documentIds?: string[];
  citationIds?: string[];
  attachmentIds?: string[];
  sharedProfileFields?: Array<"DISPLAY_NAME" | "EMAIL" | "PHONE">;
  legalIssueCategory?: string;
  contractType?: string | null;
  urgency?: string;
  userExpectedOutcome?: string;
  aiAnswerSummary?: string | null;
  aiIntent?: string | null;
  consentGranted?: boolean;
  ticket_type?: LegalTicketType;
  chat_session_id?: string | null;
  chat_message_id?: string | null;
  request_id?: string | null;
  workspace_id: string;
  document_id?: string | null;
  question: string;
  issue_fingerprint?: string | null;
  customer_note?: string | null;
}

export interface AssignLawyerRequest {
  lawyer_id: number;
  admin_note?: string | null;
  force_reassign?: boolean;
}

export interface RejectLegalTicketRequest {
  reason: string;
}

export interface CustomerTicketReplyRequest {
  message: string;
}

export interface CancelLegalTicketRequest {
  reason?: string | null;
}

export interface CloseLegalTicketRequest {
  feedback?: string | null;
}

export interface ReopenLegalTicketRequest {
  reason: string;
}

export interface LegalTicket {
  creationSource?: "AI_CHAT" | "MANUAL_FORM";
  legalIssueCategory?: string | null;
  contractType?: string | null;
  userExpectedOutcome?: string | null;
  userId?: number;
  userDisplayName?: string | null;
  userEmail?: string | null;
  userPhone?: string | null;
  selectedDocumentIds?: string[];
  sharedProfileFields?: string[];
  consentGrantedAt?: string | null;
  consentRevokedAt?: string | null;
  aiQuestion?: string | null;
  aiAnswerSummary?: string | null;
  aiIntent?: string | null;
  aiRiskLevel?: LegalTicketRiskLevel | null;
  aiConfidence?: number | null;
  ticketComplexity?: "BASIC" | "STANDARD" | "COMPLEX" | "OUT_OF_SCOPE" | null;
  classificationReason?: string | null;
  classifiedAt?: string | null;
  pricingType?: "PLAN_INCLUDED" | "PAID" | null;
  userPrice?: number | null;
  internalTicketValue?: number | null;
  quoteStatus?: "DRAFT" | "SENT" | "ACCEPTED" | "REJECTED" | "EXPIRED" | null;
  paymentStatus?: "NOT_REQUIRED" | "UNPAID" | "PENDING" | "PAID" | "FAILED" | "REFUNDED" | "PARTIALLY_REFUNDED" | null;
  quotaCycle?: string | null;
  quotaReservationStatus?: "NOT_APPLICABLE" | "PENDING" | "RESERVED" | "CONSUMED" | "RELEASED" | "RESTORED" | null;
  proposedExpertId?: number | null;
  proposedExpertName?: string | null;
  assignmentOfferedAt?: string | null;
  acceptanceDueAt?: string | null;
  acceptedAt?: string | null;
  startedAt?: string | null;
  firstResponseDueAt?: string | null;
  firstRespondedAt?: string | null;
  resolutionDueAt?: string | null;
  lastExpertActivityAt?: string | null;
  slaStatus?: "ON_TRACK" | "WAITING_FOR_USER" | "WARNING" | "OVERDUE" | "BREACHED" | "EXTENDED" | null;
  ticketCode?: string;
  title?: string;
  description?: string;
  recipientType?: TicketRecipientType;
  priority?: TicketPriority;
  conversationScope?: ConversationScope;
  sourceUserMessageId?: string;
  sourceAssistantMessageId?: string;
  focusedDocumentId?: string;
  sharedDocumentIds?: string[];
  contextSnapshot?: TicketContextSnapshot;
  id: string;
  ticket_type?: LegalTicketType;
  chat_session_id?: string;
  chat_message_id?: string;

  request_id: string;

  workspace_id?: string;
  workspace_name?: string;

  document_id?: string;
  document_name?: string;

  created_by_id?: number;
  created_by_name?: string;

  question: string;
  answer?: string;

  confidence_score?: number;
  should_suggest_ticket?: boolean;

  suggestion_type?: SuggestionType;
  suggestion_reason?: string;
  missing_information?: string;

  risk_level?: LegalTicketRiskLevel;
  legal_domain?: string;
  user_action_hint?: UserActionHint;

  status: LegalTicketStatus;

  assigned_lawyer_id?: number;
  assigned_lawyer_name?: string;

  issue_fingerprint?: string;
  customer_note?: string;

  issue_title?: string;
  issue_summary?: string;

  problematic_clause?: string;
  clause_reference?: string;
  page_number?: number;

  ai_evidence?: string;
  recommended_action?: string;

  expert_answer?: string;
  expert_internal_note?: string;
  consultation_fee?: number;
  commission_rate?: number;
  platform_fee?: number;
  expert_payout?: number;
  expert_payment_status?: "UNPAID" | "PENDING" | "PAID";
  expert_paid_at?: string;

  admin_note?: string;
  rejection_reason?: string;

  assigned_at?: string;
  resolved_at?: string;
  closed_at?: string;
  cancelled_at?: string;
  reopened_at?: string;

  created_at: string;
  updated_at: string;
}

export interface LegalTicketMessage {
  id: string;
  ticket_id: string;

  sender_id: number;
  sender_name: string;
  sender_role: string;

  content: string;
  message_type: string;

  created_at: string;

  internal_only: boolean;
  replyToMessageId?: string | null;
  editedAt?: string | null;
  attachments?: TicketAttachment[];
}

export interface TicketContextSnapshot {
  id: string;
  userQuestion: string;
  assistantAnswer?: string | null;
  conversationTitle?: string | null;
  citationSnapshotJson?: string | null;
  documentSnapshotJson?: string | null;
  selectedMessageSnapshotJson?: string | null;
  contentHash: string;
  createdAt: string;
}

export interface TicketAttachment {
  id: string;
  originalFileName: string;
  mimeType: string;
  sizeBytes: number;
  scanStatus: string;
  uploadStatus: string;
  createdAt: string;
  downloadUrl?: string;
}

export interface AttachmentPolicy {
  maxAttachmentSizeKb: number;
  maxAttachmentsPerMessage: number;
  maxAttachmentsPerTicket: number;
  allowedMimeTypes: string[];
}

export interface ConversationShare {
  id: string;
  ticketId: string;
  shareUrl: string;
  scope: ConversationScope;
  accessMode: string;
  expiresAt: string;
  revokedAt?: string | null;
  createdAt: string;
}

export interface TicketSummary {
  ticketId: string;
  confidenceScore: number | null;
  riskLevel: LegalTicketRiskLevel | null;
  suggestionType: SuggestionType | null;
  suggestionReason: string | null;
  summary: string | null;
}

export interface AdminChatHistory {
  ticketId: string;
  messages: LegalTicketMessage[];
}

export interface AdminTicketFile {
  documentId: string;
  originalFileName: string;
  fileType: string;
  fileSize: number;
  visibilityScope: string;
  uploadedAt: string;
}

File: backend/src/main/java/com/analyzer/api/entity/LegalTicket.java
package com.analyzer.api.entity;

import com.analyzer.api.enums.LegalTicketStatus;
import com.analyzer.api.enums.LegalTicketType;
import com.analyzer.api.enums.RiskLevel;
import com.analyzer.api.enums.SuggestionType;
import com.analyzer.api.enums.UserActionHint;
import com.analyzer.api.enums.ConversationScope;
import com.analyzer.api.enums.TicketPriority;
import com.analyzer.api.enums.TicketRecipientType;
import com.analyzer.api.enums.TicketCreationSource;
import com.analyzer.api.enums.TicketComplexity;
import com.analyzer.api.enums.TicketPricingType;
import com.analyzer.api.enums.TicketQuoteStatus;
import com.analyzer.api.enums.TicketPaymentStatus;
import com.analyzer.api.enums.TicketQuotaReservationStatus;
import com.analyzer.api.enums.TicketSlaStatus;
import com.analyzer.api.enums.FailureResponsibleParty;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;

import java.time.LocalDateTime;
import java.math.BigDecimal;
import java.util.UUID;

@Entity
@Table(
    name = "legal_tickets",
    uniqueConstraints = {
        @UniqueConstraint(
            name = "uk_legal_ticket_request_creator",
            columnNames = {"request_id", "created_by_id"}
        )
    }
)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
public class LegalTicket extends BaseEntity {

    @Id
    private String id;

    @Column(name = "ticket_code", unique = true)
    private String ticketCode;

    @Column(name = "request_id", nullable = false)
    private String requestId;

    @Column(name = "issue_fingerprint")
    private String issueFingerprint;

    @Column(name = "customer_note", columnDefinition = "TEXT")
    private String customerNote;

    @Enumerated(EnumType.STRING)
    @Column(name = "ticket_type")
    private LegalTicketType ticketType;

    @Enumerated(EnumType.STRING)
    @Column(name = "creation_source", nullable = false)
    private TicketCreationSource creationSource;

    @Column(name = "legal_issue_category")
    private String legalIssueCategory;

    @Column(name = "contract_type")
    private String contractType;

    @Column(name = "user_expected_outcome", columnDefinition = "TEXT")
    private String userExpectedOutcome;

    @Column(name = "shared_profile_fields_json", columnDefinition = "TEXT")
    private String sharedProfileFieldsJson;

    @Column(name = "consent_granted_at")
    private LocalDateTime consentGrantedAt;

    @Column(name = "consent_revoked_at")
    private LocalDateTime consentRevokedAt;

    @Column(name = "related_chat_session_id")
    private String relatedChatSessionId;

    @Column(name = "related_chat_message_id")
    private String relatedChatMessageId;

    @Column(name = "source_user_message_id")
    private String sourceUserMessageId;

    @Column(name = "source_assistant_message_id")
    private String sourceAssistantMessageId;

    @Enumerated(EnumType.STRING)
    @Column(name = "recipient_type")
    private TicketRecipientType recipientType;

    @Column(name = "title")
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "priority")
    private TicketPriority priority;

    @Enumerated(EnumType.STRING)
    @Column(name = "conversation_scope")
    private ConversationScope conversationScope;

    @Column(name = "shared_document_ids_json", columnDefinition = "TEXT")
    private String sharedDocumentIdsJson;

    @Column(name = "focused_document_id")
    private String focusedDocumentId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_id", nullable = false)
    private User createdBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "workspace_id")
    private Workspace workspace;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "document_id")
    private Document document;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String question;

    @Column(columnDefinition = "TEXT")
    private String answer;

    @Column(name = "confidence_score")
    private Double confidenceScore;

    @Column(name = "should_suggest_ticket")
    private Boolean shouldSuggestTicket;

    @Enumerated(EnumType.STRING)
    @Column(name = "suggestion_type")
    private SuggestionType suggestionType;

    @Column(name = "suggestion_reason", columnDefinition = "TEXT")
    private String suggestionReason;

    @Column(name = "missing_information", columnDefinition = "TEXT")
    private String missingInformation;

    @Enumerated(EnumType.STRING)
    @Column(name = "risk_level")
    private RiskLevel riskLevel;

    @Column(name = "legal_domain")
    private String legalDomain;

    @Enumerated(EnumType.STRING)
    @Column(name = "user_action_hint")
    private UserActionHint userActionHint;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private LegalTicketStatus status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_lawyer_id")
    private User assignedLawyer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "proposed_expert_id")
    private User proposedExpert;

    @Enumerated(EnumType.STRING)
    @Column(name = "ticket_complexity")
    private TicketComplexity ticketComplexity;

    @Column(name = "classification_reason", columnDefinition = "TEXT")
    private String classificationReason;

    @Column(name = "classified_at")
    private LocalDateTime classifiedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "classified_by_id")
    private User classifiedBy;

    @Enumerated(EnumType.STRING)
    @Column(name = "pricing_type")
    private TicketPricingType pricingType;

    @Column(name = "user_price", precision = 19, scale = 2)
    private BigDecimal userPrice;

    @Column(name = "internal_ticket_value", precision = 19, scale = 2)
    private BigDecimal internalTicketValue;

    @Enumerated(EnumType.STRING)
    @Column(name = "quote_status")
    private TicketQuoteStatus quoteStatus;

    @Enumerated(EnumType.STRING)
    @Column(name = "customer_payment_status")
    private TicketPaymentStatus customerPaymentStatus;

    @Column(name = "customer_payment_reference")
    private String customerPaymentReference;

    @Column(name = "quota_cycle")
    private String quotaCycle;

    @Enumerated(EnumType.STRING)
    @Column(name = "quota_reservation_status")
    private TicketQuotaReservationStatus quotaReservationStatus;

    @Column(name = "assignment_offered_at")
    private LocalDateTime assignmentOfferedAt;

    @Column(name = "acceptance_due_at")
    private LocalDateTime acceptanceDueAt;

    @Column(name = "accepted_at")
    private LocalDateTime acceptedAt;

    @Column(name = "started_at")
    private LocalDateTime startedAt;

    @Column(name = "first_response_due_at")
    private LocalDateTime firstResponseDueAt;

    @Column(name = "first_responded_at")
    private LocalDateTime firstRespondedAt;

    @Column(name = "resolution_due_at")
    private LocalDateTime resolutionDueAt;

    @Column(name = "completed_at")
    private LocalDateTime completedAt;

    @Column(name = "last_expert_activity_at")
    private LocalDateTime lastExpertActivityAt;

    @Enumerated(EnumType.STRING)
    @Column(name = "sla_status")
    private TicketSlaStatus slaStatus;

    @Column(name = "paused_at")
    private LocalDateTime pausedAt;

    @Column(name = "total_paused_duration_seconds")
    private Long totalPausedDurationSeconds;

    @Column(name = "extension_reason", columnDefinition = "TEXT")
    private String extensionReason;

    @Enumerated(EnumType.STRING)
    @Column(name = "failure_responsible_party")
    private FailureResponsibleParty failureResponsibleParty;

    @Column(name = "previous_expert_id")
    private Long previousExpertId;

    @Column(name = "reassignment_reason", columnDefinition = "TEXT")
    private String reassignmentReason;

    @Column(name = "reassigned_at")
    private LocalDateTime reassignedAt;

    @Column(name = "reassigned_by_id")
    private Long reassignedById;

    @Column(name = "completion_percent")
    private Integer completionPercent;

    @Column(name = "approved_partial_payout", precision = 19, scale = 2)
    private BigDecimal approvedPartialPayout;

    @Column(name = "contribution_note", columnDefinition = "TEXT")
    private String contributionNote;

    // Snapshot fields
    @Column(name = "issue_title")
    private String issueTitle;

    @Column(name = "issue_summary", columnDefinition = "TEXT")
    private String issueSummary;

    @Column(name = "problematic_clause", columnDefinition = "TEXT")
    private String problematicClause;

    @Column(name = "clause_reference")
    private String clauseReference;

    @Column(name = "page_number")
    private Integer pageNumber;

    @Column(name = "ai_evidence", columnDefinition = "TEXT")
    private String aiEvidence;

    @Column(name = "recommended_action", columnDefinition = "TEXT")
    private String recommendedAction;

    // Resolution and notes
    @Column(name = "expert_answer", columnDefinition = "TEXT")
    private String expertAnswer;

    @Column(name = "expert_internal_note", columnDefinition = "TEXT")
    private String expertInternalNote;

    @Column(name = "admin_note", columnDefinition = "TEXT")
    private String adminNote;

    @Column(name = "rejection_reason", columnDefinition = "TEXT")
    private String rejectionReason;

    // Audit timestamps
    @Column(name = "assigned_at")
    private LocalDateTime assignedAt;

    @Column(name = "resolved_at")
    private LocalDateTime resolvedAt;

    @Column(name = "closed_at")
    private LocalDateTime closedAt;

    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;

    @Column(name = "reopened_at")
    private LocalDateTime reopenedAt;

    @Column(name = "close_reason", columnDefinition = "TEXT")
    private String closeReason;

    @Column(name = "last_customer_message_at")
    private LocalDateTime lastCustomerMessageAt;

    @Column(name = "last_lawyer_message_at")
    private LocalDateTime lastLawyerMessageAt;

    @Column(name = "consultation_fee", precision = 19, scale = 2)
    @Builder.Default
    private BigDecimal consultationFee = BigDecimal.ZERO;

    @Enumerated(EnumType.STRING)
    @Column(name = "expert_payment_status", nullable = false)
    @Builder.Default
    private com.analyzer.api.enums.ExpertPaymentStatus expertPaymentStatus = com.analyzer.api.enums.ExpertPaymentStatus.UNPAID;

    @Column(name = "expert_paid_at")
    private LocalDateTime expertPaidAt;

    @Column(name = "commission_rate", precision = 5, scale = 4)
    private BigDecimal commissionRate;

    @Column(name = "platform_fee", precision = 19, scale = 2)
    private BigDecimal platformFee;

    @Column(name = "expert_payout", precision = 19, scale = 2)
    private BigDecimal expertPayout;

    @Column(nullable = false)
    @Builder.Default
    private boolean deleted = false;

    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    @Version
    private Long version;

    @PrePersist
    protected void onCreate() {
        if (this.id == null || this.id.isBlank()) {
            this.id = "ticket_" + UUID.randomUUID().toString().replace("-", "");
        }
        if (this.ticketCode == null || this.ticketCode.isBlank()) {
            this.ticketCode = "TKT-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
        }
        if (this.priority == null) this.priority = TicketPriority.NORMAL;
        if (this.creationSource == null) this.creationSource = TicketCreationSource.MANUAL_FORM;
        if (this.quoteStatus == null) this.quoteStatus = TicketQuoteStatus.DRAFT;
        if (this.customerPaymentStatus == null) this.customerPaymentStatus = TicketPaymentStatus.UNPAID;
        if (this.quotaReservationStatus == null) this.quotaReservationStatus = TicketQuotaReservationStatus.PENDING;
        if (this.slaStatus == null) this.slaStatus = TicketSlaStatus.ON_TRACK;
        if (this.totalPausedDurationSeconds == null) this.totalPausedDurationSeconds = 0L;
        if (this.recipientType == null) this.recipientType = TicketRecipientType.EXPERT;
        if (this.conversationScope == null) this.conversationScope = ConversationScope.SELECTED_RESPONSE;
        this.deleted = false;
    }
}
